[SYCL][ESIMD][EMU] Marking ESIMD kernels for esimd_emulator backend

dongkyunahn-intel opened the pull request:
This PR is to mark ESIMD kernels for esimd_emulator backend with `UNSUPPORTED` or `XFAIL` for future feature bring-up and debug.

dongkyunahn-intel committed:
Duplicating ESIMD samples for esimd_emulator

- 'vadd_1d' and 'sycl_esimd_mix' are duplicated to
'esimd_check_vc_codegen.cpp' and 'sycl_esimd_mix_check_build_opts.cpp'
so that original kernels can run with esimd_emulator without 'CHECK'
commands and duplicated kernels can run with opencl/level_zero
backends with 'CHECK'

dongkyunahn-intel committed:
Adding 'esimd_emualtor' as supported SYCL backend

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+    C[i] = 0.0f;
+  }
+
+  try {
```
wouldn't an empty kernel run via `single_task` (or `parallel_for` if single_task is not supported yet) suffice here?

dongkyunahn-intel replied:
I'm not sure if it is right to modify this test as it is duplicate of `vadd_1d.cpp`. Original `vadd_1d.cpp` is changed to remove `CHECK` commands for `piProgramBuild` and to mark `XFAIL: esimd_emulator`. This test has those `CHECK` commands and `UNSUPPORTED: esimd_emulator` as `esimd_emulator` does not support online compilation

Do you want this test to be removed like `sycl_esimd_mix_check_build_opts.cpp` and original `vadd_1d.cpp` to have `UNSUPPORTED: esimd_emulator`?

kbobrovs replied:
see suggested changes

dongkyunahn-intel replied:
Suggestions applied.

dongkyunahn-intel commented:
This PR will be merged after ESIMD_EMULATOR support is enabled by default in intel/llvm - https://github.com/intel/llvm/pull/5058.

kbobrovs commented on SYCL/ESIMD/sycl_esimd_mix.cpp:
```diff
@@ -10,9 +10,11 @@
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// TODO/FIXME: esimd_emulator support - timeout
+// XFAIL: esimd_emulator
```
Changes in this test completely defeat the test's purpose. I think for now, we should just add `UNSUPPORTED: esimd_emulator`, and leave the rest as is.
Eventually, we might need to figure out a way to execute non-ESIMD kernels via IGC and ESIMD kernels - via esimd_emulator.

dongkyunahn-intel replied:
Will do.

kbobrovs commented on SYCL/ESIMD/sycl_esimd_mix_check_build_opts.cpp:
```diff
@@ -0,0 +1,148 @@
+//==--- sycl_esimd_mix_check_build_opts.cpp - DPC++ ESIMD on-device test ---==//
```
according to my previous comment, this test is not needed

dongkyunahn-intel replied:
Will remove.

dongkyunahn-intel replied:
Done.

vladimirlaz commented on SYCL/lit.cfg.py:
```diff
@@ -189,7 +189,8 @@
                      'hip',
-                     'level_zero']
+                     'level_zero',
+                     'esimd_emulator']
```
If you have no objections, could we use intel/llvm-test-suite#752 PR to enable esimd_emulator in LIT infra. This PR can be used to have fixes to tests.

dongkyunahn-intel replied:
I have no objections - more than that I agree. I'll revert this change.

dongkyunahn-intel committed:
Revert "Adding 'esimd_emualtor' as supported SYCL backend"

This reverts commit c5b6567166c907d166db65cfa1bc80a6a68baa85.

dongkyunahn-intel committed:
Reverting changes in 'SYCL/ESIMD/sycl_esimd_mix.cpp'

- With 'UNSUPPORTED: esimd_emulator' and comment

dongkyunahn-intel committed:
Fixing typo in comment for SYCL/ESIMD/vadd_1d.cpp

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+using namespace cl::sycl;
+
+int main(void) {
+  constexpr unsigned Size = 1024 * 128;
+  constexpr unsigned VL = 16;
+
+  float *A = new float[Size];
+  float *B = new float[Size];
+  float *C = new float[Size];
+
+  for (unsigned i = 0; i < Size; ++i) {
+    A[i] = B[i] = i;
+    C[i] = 0.0f;
+  }
+
```
```suggestion
```

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  try {
+    buffer<float, 1> bufa(A, range<1>(Size));
+    buffer<float, 1> bufb(B, range<1>(Size));
+    buffer<float, 1> bufc(C, range<1>(Size));
+
+    // We need that many workgroups
+    range<1> GlobalRange{Size / VL};
+
+    // We need that many threads in each group
+    range<1> LocalRange{1};
+
```
```suggestion
```

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+    std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+    auto e = q.submit([&](handler &cgh) {
+      auto PA = bufa.get_access<access::mode::read>(cgh);
+      auto PB = bufb.get_access<access::mode::read>(cgh);
+      auto PC = bufc.get_access<access::mode::write>(cgh);
+      cgh.parallel_for<class Test>(
+          GlobalRange * LocalRange, [=](id<1> i) SYCL_ESIMD_KERNEL {
+            using namespace sycl::ext::intel::experimental::esimd;
+            unsigned int offset = i * VL * sizeof(float);
+            simd<float, VL> va;
+            va.copy_from(PA, offset);
+            simd<float, VL> vb;
+            vb.copy_from(PB, offset);
+            simd<float, VL> vc = va + vb;
+            vc.copy_to(PC, offset);
+          });
```
```suggestion
  cgh.single_task<class Test>([]{});
```

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+
+    delete[] A;
+    delete[] B;
+    delete[] C;
+    return 1;
```
```suggestion
    return 1;
```

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,104 @@
+    delete[] C;
+    return 1;
+  }
+
+  int err_cnt = 0;
+
+  for (unsigned i = 0; i < Size; ++i) {
+    if (A[i] + B[i] != C[i]) {
+      if (++err_cnt < 10) {
+        std::cout << "failed at index " << i << ", " << C[i] << " != " << A[i]
+                  << " + " << B[i] << "\n";
+      }
+    }
+  }
+  if (err_cnt > 0) {
+    std::cout << "  pass rate: "
+              << ((float)(Size - err_cnt) / (float)Size) * 100.0f << "% ("
+              << (Size - err_cnt) << "/" << Size << ")\n";
+  }
+
+  delete[] A;
+  delete[] B;
+  delete[] C;
+
+  std::cout << (err_cnt > 0 ? "FAILED\n" : "Passed\n");
+  return err_cnt > 0 ? 1 : 0;
```
```suggestion
  return 0;
```

dongkyunahn-intel replied:
@kbobrovs , `esimd_check_vc_codegen.cpp` fails after being simplified with `single_task` - for both open_cl and level_zero. Maybe it has to be reverted.

```
[2022-01-27T05:52:29.296Z] /netbatch/donb02499_00/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm-test-suite/SYCL/ESIMD/esimd_check_vc_codegen.cpp:41:11: error: CHECK: expected string not found in input
[2022-01-27T05:52:29.296Z] // CHECK: <const char *>: {{.*}}-vc-codegen
[2022-01-27T05:52:29.296Z]           ^
[2022-01-27T05:52:29.296Z] <stdin>:415:2: note: scanning from here
[2022-01-27T05:52:29.296Z]  <unknown> : 0x7056b0
[2022-01-27T05:52:29.296Z]  ^
[2022-01-27T05:52:29.296Z] <stdin>:418:2: note: possible intended match here
[2022-01-27T05:52:29.296Z]  <const char *>: 
[2022-01-27T05:52:29.296Z]  ^
[2022-01-27T05:52:29.296Z] 
```

dongkyunahn-intel committed:
Apply suggestions from code review

Co-authored-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

dongkyunahn-intel committed:
Marking ESIMD kernels 'UNSUPPORTED' for esimd_emulator backend

dongkyunahn-intel committed:
clang-format error fix

dongkyunahn-intel committed:
Merge branch 'intel' of https://github.com/intel/llvm-test-suite into invoke_esimd_emulator_check_in

dongkyunahn-intel committed:
XFAIL markups replacing UNSUPPORTED

- Unimplemented features
- Outdated memory intrinsic implementations

Matrix size limitation - matrix_transpose_glb.cpp

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,44 @@
+    std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+    auto e = q.submit([&](handler &cgh) {
+  cgh.single_task<class Test>([]{});
```
please fix formatting

dongkyunahn-intel replied:
Done.

kbobrovs commented on SYCL/ESIMD/vadd_1d.cpp:
```diff
@@ -7,8 +7,10 @@
 //===----------------------------------------------------------------------===//
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// TODO: esimd_emulator support - enable __esimd_oword_ld_unaligned
```
```suggestion
```

kbobrovs commented on SYCL/ESIMD/esimd_check_vc_codegen.cpp:
```diff
@@ -0,0 +1,44 @@
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+
```
```suggestion
```

dongkyunahn-intel replied:
Done.

kbobrovs commented on SYCL/ESIMD/vadd_1d.cpp:
```diff
@@ -7,8 +7,10 @@
 //===----------------------------------------------------------------------===//
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// TODO: esimd_emulator support - enable __esimd_oword_ld_unaligned
```
```suggestion
// TODO: esimd_emulator fails due to unimplemented __esimd_oword_ld_unaligned
```

dongkyunahn-intel replied:
Done.

kbobrovs commented on SYCL/ESIMD/vadd_1d.cpp:
```diff
@@ -97,6 +99,5 @@ int main(void) {
-// CHECK: ---> piProgramBuild(
-// CHECK: <const char *>: {{.*}}-vc-codegen
-// CHECK: ) ---> pi_result : PI_SUCCESS
+// 'CHECK' commands for checking 'vc-codegen' are moved to
+// 'esimd_check_vc_codegen.cpp'
```
```suggestion
```

dongkyunahn-intel replied:
Done.

dongkyunahn-intel committed:
clang-format fix

- Missing XFAIL markup replacing UNSUPPORTED

kbobrovs commented on SYCL/ESIMD/histogram_256_slm_spec_2020.cpp:
```diff
@@ -1,6 +1,7 @@
 // TODO enable on Windows
 // REQUIRES: linux && gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
TODO?

kbobrovs commented on SYCL/ESIMD/mandelbrot/mandelbrot_spec.cpp:
```diff
@@ -9,6 +9,7 @@
 // TODO enable on Windows
 // REQUIRES: linux && gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
TODO?

kbobrovs commented on SYCL/ESIMD/regression/complex-lib-lin.cpp:
```diff
@@ -7,6 +7,7 @@
 //
 // REQUIRES: linux,gpu
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
TODO?

kbobrovs commented on SYCL/ESIMD/regression/globals.cpp:
```diff
@@ -1,5 +1,7 @@
 // REQUIRES: gpu
 // UNSUPPORTED: cuda || hip
+// Sub-group is not supported
```
```suggestion
// TODO Sub-group is not supported
```

kbobrovs commented on SYCL/ESIMD/spec_const/spec_const_bool.cpp:
```diff
@@ -9,6 +9,7 @@
 // RUN: %clangxx -fsycl -I%S/.. %s -o %t.out
 // RUN: %GPU_RUN_PLACEHOLDER %t.out
 // UNSUPPORTED: cuda || hip
+// XFAIL: esimd_emulator
```
TODO?  here and other places. We should eventually support spec constants in the emulator.

dongkyunahn-intel replied:
Like I showed in messenger, online_compiler check fails for all spec* tests - even if it is not required. To whom should I ask about this issue? (Someone in SYCL_RT?)

dongkyunahn-intel committed:
'esimd_emulator' as deprecated name for 'ext_intel_esimd_emulator'

- as 'esimd_emulator' is substring of 'ext_intel_esimd_emulator'

dongkyunahn-intel committed:
Missing TODO comments for esimd_emulator support markup

dongkyunahn-intel committed:
Adding dummy codes to invoke piProgramBuild for vc_codegen

dongkyunahn-intel committed:
Removing HOST_RUN_PLACEHOLDER for ESIMD Kernels

- Host backend and esimd_emulator backend share same header file for
memory intrinsic support - memory_intrin.h

- This sharing prevents the same header file from supporting both
backend types as some intrinsic support (e.g. __esimd_media_ld/st) are
different and they are chosen during kernel compilation - while
backends are chosen during kernel execution

- Therefore, tests invoking those implementations must not contain
'HOST_RUN_PLACEHOLDER' as the implementations require esimd_emulator
backend

dongkyunahn-intel committed:
Recovering 'esimd_check_vc_codegen.cpp' to its initial import version

- As 'CHECK' fails for 'piProgramBuild' PI_API call

dongkyunahn-intel committed:
Merge branch 'intel' of https://github.com/intel/llvm-test-suite into invoke_esimd_emulator_check_in

dongkyunahn-intel commented:
`esimd_check_vc_codegen.cpp` will be revised to minimized test as @kbobrovs suggested in a separate PR.

kbobrovs reviewed (dismissed):
LGTM, but please address the vc_code_gen test comment at your earliest convenience.

dongkyunahn-intel committed:
Merge branch 'intel' of https://github.com/intel/llvm-test-suite into invoke_esimd_emulator_check_in

dongkyunahn-intel committed:
Changing 'UNSUPPORTED' to 'XFAIL' for esimd_emulator backend

- For tests failed due to unimplemented features such as 'half' type

dongkyunahn-intel committed:
Marking 'XFAIL' for newly added tests

dongkyunahn-intel committed:
Revert "Removing HOST_RUN_PLACEHOLDER for ESIMD Kernels"

This reverts commit 697e3bd324a32130105939f7ba2945d067216a9c.

dongkyunahn-intel committed:
Revert "'esimd_emulator' as deprecated name for 'ext_intel_esimd_emulator'"

This reverts commit 092f7c1e3f74d5229c2b7d7b649b4f7b0276965c.
